### ISSUE TITLE
How to correctly handle my pd3dDevice to be used for loading textures throughout runtime?

### ISSUE THREAD

**Issue by abumettleq** [ocornut/imgui#7347]
### Version/Branch of Dear ImGui:

Version 1.90.3, Branch: Help Needed

### Back-ends:

imgui_impl_dx9.cpp + imgui_impl_win32.cpp

### Compiler, OS:

Windows 11 + MSVC 2022 + C++17

### Details:

How do I initialize my pd3dDevice and keep it alive (non-nullptr) during the lifetime of the process?

I have tried the following in the code below.

1. I have defined my ImageHelper
```
static LPDIRECT3DDEVICE9 pd3dDevice;

void LoadTextureFromFile(const char* filename, PDIRECT3DTEXTURE9* out_texture)
{
	// Load texture from disk
	PDIRECT3DTEXTURE9 texture;
	HRESULT hr = D3DXCreateTextureFromFileA(pd3dDevice, filename, &texture);
	if (hr != S_OK)
	{
		if(!pd3dDevice) MessageBox(NULL, "pd3dDevice died.", "pd3dDevice", MB_OK);

		MessageBox(NULL,"Failed to load slot texture.", filename, MB_OK);
		return;
	}

	// Retrieve description of the texture surface so we can access its size
	D3DSURFACE_DESC my_image_desc;
	texture->GetLevelDesc(0, &my_image_desc);
	*out_texture = texture;
}
```

2. At first frame, the font will be loaded along with creation of device objects (once only).
```
bool isFontLoaded = false;
void AsdaStoryD3D9HookHandler::HandleNewFrame()
{
	if (currentProxy == nullptr)
		return;


	if (!isFontLoaded)
	{
		ImGui_ImplDX9_InvalidateDeviceObjects();
		//TextureFactory::Instance()->ReleaseTextures();

		ImGui::GetIO().Fonts->AddFontFromFileTTF(FONT_PATH(AGAR_FONT), 13);
		ImGui::GetIO().Fonts->AddFontFromFileTTF(FONT_PATH(AGAR_FONT), 14);
		ImGui::GetIO().Fonts->AddFontFromFileTTF(FONT_PATH(AGAR_FONT), 15);
		ImGui::GetIO().Fonts->AddFontFromFileTTF(FONT_PATH(AGAR_FONT), 16);
		ImGui::GetIO().Fonts->AddFontFromFileTTF(FONT_PATH(AGAR_FONT), 17);

		ImGui_ImplDX9_CreateDeviceObjects();

		AsdaExtension::UI::Avatar::LoadAvatarSlotsTextures();

		isFontLoaded = true;
	}

	//AsdaExtension::UI::DailyBonus::DailyBonusWindow();
	AsdaExtension::UI::Avatar::AvatarWindow();
}
```

3. Initialized my pd3dDevice at time of creation of device objects.
```
bool ImGui_ImplDX9_CreateDeviceObjects()
{
    ImGui_ImplDX9_Data* bd = ImGui_ImplDX9_GetBackendData();
    if (!bd || !bd->pd3dDevice)
        return false;
    if (!ImGui_ImplDX9_CreateFontsTexture())
        return false;

    UI::ImageHelper::pd3dDevice = bd->pd3dDevice;
    return true;
}
```

4. Loaded the the textures.
```
void LoadAvatarSlotsTextures()
{
	if (!pd3dDevice)
	{
		LoadTextureFromFile("eIcon/chest.bmp", &chest);
		LoadTextureFromFile("eIcon/guantlet.bmp", &guantlet);
		LoadTextureFromFile("eIcon/helmet.bmp", &helmet);
		LoadTextureFromFile("eIcon/necklace.bmp", &necklace);
		LoadTextureFromFile("eIcon/pants.bmp", &pants);
		LoadTextureFromFile("eIcon/ring.bmp", &ring);
		LoadTextureFromFile("eIcon/shield.bmp", &shield);
		LoadTextureFromFile("eIcon/shoes.bmp", &shoes);
		LoadTextureFromFile("eIcon/weapon.bmp", &weapon);
	}
	else MessageBox(NULL, "pd3dDevice is null!, failed to load slots.", "DEBUG", MB_OK);
}
```

### Screenshots/Video:

1. First Debug Message:
![image](https://github.com/ocornut/imgui/assets/84507882/1772045c-c4e1-4408-abcf-004225b050d6)

2. Second Debug Message:
![image](https://github.com/ocornut/imgui/assets/84507882/6c37312e-8e6a-4f1f-9a91-d41e611e17b1)

same for each texture file... and so on.

### Minimal, Complete and Verifiable Example code:

```cpp
// Here's some code anyone can copy and paste to reproduce your issue
ImGui::Begin("Example Bug");
MoreCodeToExplainMyIssue();
ImGui::End();
```

**Comment by abumettleq** [comment]
Solved it.

```
bool ImGui_ImplDX9_Init(IDirect3DDevice9* device)
{
    ImGuiIO& io = ImGui::GetIO();
    IM_ASSERT(io.BackendRendererUserData == nullptr && "Already initialized a renderer backend!");

    // Setup backend capabilities flags
    ImGui_ImplDX9_Data* bd = IM_NEW(ImGui_ImplDX9_Data)();
    io.BackendRendererUserData = (void*)bd;
    io.BackendRendererName = "imgui_impl_dx9";
    io.BackendFlags |= ImGuiBackendFlags_RendererHasVtxOffset;  // We can honor the ImDrawCmd::VtxOffset field, allowing for large meshes.

    bd->pd3dDevice = device;
    AsdaExtension::UI::ImageHelper::TextureLoader::Instance()->SetDevice(device);
    bd->pd3dDevice->AddRef();

    return true;
}
```

**Comment by GamingMinds-DanielC** [comment]
Wha do you not just set the device into your texture loader where you perform the call to `ImGui_ImplDX9_Init()`? Needlessly modifying the backend will only limit your ability to update the library.

On second thought, Your texture loader just holds an additional reference on top of the ImGui backend and the one your application should hold. If that actually prevents the device from being released, that's a sign that something is seriously wrong somewhere in your application. Without the texture loader, you are releasing more references than you acquire. The additional reference in the texture loader may be a functioning workaround, but it doesn't solve the core issue.

**Comment by abumettleq** [comment]
Oh, If I got you right, is this how am I supposed to initialize my `pd3dDevice`?
```
IMGUI_CHECKVERSION();
ImGui::CreateContext();
ImGuiIO& io = ImGui::GetIO(); (void)io;
io.IniFilename = NULL;
//io.ConfigFlags |= ImGuiConfigFlags_NavEnableKeyboard;  // Enable Keyboard Controls
io.ConfigFlags |= ImGuiConfigFlags_NoMouseCursorChange;
ImGui_ImplWin32_Init(hFocusWindow);
ImGui_ImplDX9_Init(_Std_CreateDevice_Device);
AsdaExtension::UI::ImageHelper::TextureLoader::Instance()->SetDevice(_Std_CreateDevice_Device);
```

to note that I have change my `ImageHelper` namespace as following,
```
#include "ImageHelper.hpp"
#include <string>


namespace AsdaExtension::UI::ImageHelper
{
	TextureLoader* TextureLoader::m_textureLoader = nullptr;

	TextureLoader* TextureLoader::Instance()
	{
		if (m_textureLoader == nullptr)
		{
			m_textureLoader = new TextureLoader();
		}
		return m_textureLoader;
	}

    void UI::ImageHelper::TextureLoader::SetDevice(LPDIRECT3DDEVICE9 _pd3dDevice)
    {
        this->pd3dDevice = _pd3dDevice;
    }


    void TextureLoader::LoadTextureFromZipFile(const std::string btzPath, const std::string& filename, PDIRECT3DTEXTURE9* out_texture) {
        if (!pd3dDevice) {
            MessageBox(NULL, "Direct3D device not initialized.", "Error", MB_OK);
            return;
        }
        ziputils::unzipper zipFile;
        zipFile.open(btzPath.c_str());

        if (zipFile.openEntry(filename.c_str()) == false) {
            MessageBox(NULL, "Couldn't find file in the zip archive.", filename.c_str(), MB_OK);
            zipFile.closeEntry();
            zipFile.close();
            return;
        }

        Buffer* pBuffer;
        zipFile >> pBuffer;

        if (!pBuffer->GetMemory().lpAddress)
        {
            MessageBox(NULL, "Failed to read image buffer.", filename.c_str(), MB_OK);
            zipFile.closeEntry();
            zipFile.close();
            return;
        }

        // Load texture from memory buffer
        PDIRECT3DTEXTURE9 texture;
        HRESULT hr = D3DXCreateTextureFromFileInMemory(pd3dDevice, &pBuffer->GetMemory().lpAddress[0], pBuffer->GetOffset(), &texture);
        if (hr != S_OK) {
            MessageBox(NULL, "Failed to create texture from memory buffer.", filename.c_str(), MB_OK);
            zipFile.closeEntry();
            zipFile.close();
            return;
        }

        *out_texture = texture;
        zipFile.closeEntry();
        zipFile.close();
    }
}
```

**Comment by abumettleq** [comment]
After taking a look at the examples folder `[example_win32_directx9]`, I have understood the following,

```
static LPDIRECT3DDEVICE9        g_pd3dDevice = nullptr; // we declare a LPDIRECT3DDEVICE9

// in main,
ImGui_ImplDX9_Init(g_pd3dDevice); // we pass the g_pd3dDevice by reference, to be initialized there
```

In my case , it would be like this,
```
// d3d9hook.cpp
ImGui_ImplDX9_Init(_Std_CreateDevice_Device);
if(_Std_CreateDevice_Device != nullptr)
{
	AsdaExtension::UI::ImageHelper::TextureLoader::Instance()->SetDevice(&_Std_CreateDevice_Device);
}
```

and my `TextureLoader` class would be something like this,
```
class TextureLoader
{
private:
	LPDIRECT3DDEVICE9* pd3dDevice = NULL;
	
	static TextureLoader* m_textureLoader;
public:
	static TextureLoader* Instance();
	void SetDevice(LPDIRECT3DDEVICE9* _pd3dDevice);
	void LoadTextureFromZipFile(const std::string btzPath, const std::string& filename, PDIRECT3DTEXTURE9* out_texture);
};
```

**Comment by GamingMinds-DanielC** [comment]
Your texture loader now takes a pointer to a pointer (the LP... typedef isalready a pointer), that's not necessary. You need this construct only when outputting to a pointer, as you do in the out_texture parameter in your LoadTextureFromZipFile.

**Comment by abumettleq** [comment]
Is alright now?

1. Assigning
```
ImGui_ImplDX9_Init(_Std_CreateDevice_Device);
if(_Std_CreateDevice_Device != nullptr)
{
	AsdaExtension::UI::TextureLoader::Instance()->SetDevice(_Std_CreateDevice_Device);
}
```

2.  Setter Function
```
void UI::TextureLoader::SetDevice(LPDIRECT3DDEVICE9 _pd3dDevice)
{
    this->pd3dDevice = _pd3dDevice;
}
```

3. Call for Loading Textures
```
void TextureStorage::LoadAllTex()
{
	if (TextureLoader::Instance() != nullptr)
	{

		TextureLoader::Instance()->LoadTextureFromZipFile(eIconBtz, "avatar_chest.bmp", &avatar_chest);
		TextureLoader::Instance()->LoadTextureFromZipFile(eIconBtz, "avatar_guantlet.bmp", &avatar_guantlet);
		TextureLoader::Instance()->LoadTextureFromZipFile(eIconBtz, "avatar_head.bmp", &avatar_head);
		TextureLoader::Instance()->LoadTextureFromZipFile(eIconBtz, "avatar_pants.bmp", &avatar_pants);
		TextureLoader::Instance()->LoadTextureFromZipFile(eIconBtz, "avatar_shoes.bmp", &avatar_shoes);
		TextureLoader::Instance()->LoadTextureFromZipFile(eIconBtz, "avatar_weapon.bmp", &avatar_weapon);
		TextureLoader::Instance()->LoadTextureFromZipFile(eIconBtz, "avatar_cloack.bmp", &avatar_cloack);
		TextureLoader::Instance()->LoadTextureFromZipFile(eIconBtz, "avatar_head_acc.bmp", &avatar_head_acc);
		TextureLoader::Instance()->LoadTextureFromZipFile(eIconBtz, "avatar_window_ico.bmp", &avatar_window_ico);

		TextureLoader::Instance()->LoadTextureFromZipFile(eIconBtz, "treasure_box.bmp", &treasure_box);
		TextureLoader::Instance()->LoadTextureFromZipFile(eIconBtz, "mes_present.png", &mes_present);
		TextureLoader::Instance()->LoadTextureFromZipFile(eIconBtz, "wheeltaget_once.bmp", &wheeltaget_once);
	}
	else MessageBox(NULL, "pd3dDevice is null!, failed to load slots.", "DEBUG", MB_OK);
}
```

4. Load from Zip
```
void TextureLoader::LoadTextureFromZipFile(const std::string btzPath, const std::string& filename, PDIRECT3DTEXTURE9* out_texture) {
    if (!pd3dDevice) {
        MessageBox(NULL, "Direct3D device not initialized.", "Error", MB_OK);
        return;
    }
    ziputils::unzipper zipFile;
    zipFile.open(btzPath.c_str());

    if (zipFile.openEntry(filename.c_str()) == false) {
        MessageBox(NULL, "Couldn't find file in the zip archive.", filename.c_str(), MB_OK);
        zipFile.closeEntry();
        zipFile.close();
        return;
    }

    Buffer* pBuffer;
    zipFile >> pBuffer;

    if (!pBuffer->GetMemory().lpAddress)
    {
        MessageBox(NULL, "Failed to read image buffer.", filename.c_str(), MB_OK);
        zipFile.closeEntry();
        zipFile.close();
        return;
    }

    // Load texture from memory buffer
    PDIRECT3DTEXTURE9 texture;
    HRESULT hr = D3DXCreateTextureFromFileInMemory(pd3dDevice, &pBuffer->GetMemory().lpAddress[0], pBuffer->GetOffset(), &texture);
    if (hr != S_OK) {
        MessageBox(NULL, "Failed to create texture from memory buffer.", filename.c_str(), MB_OK);
        zipFile.closeEntry();
        zipFile.close();
        return;
    }

    *out_texture = texture;
    zipFile.closeEntry();
    zipFile.close();
}
```

**Comment by ocornut** [comment]
Feel free to discuss this here but I'm closing the thread at it is essentially off-topic.

**Comment by GamingMinds-DanielC** [comment]
> Feel free to discuss this here but I'm closing the thread at it is essentially off-topic.

I don't want to take this much further as well for the same reason, just a few closing thoughts...

> Is alright now?

Better in that aspect, but lots of room for improvement. I don't know your codebase, but some general hints...

If you hold a pointer to a device or device object, you should also hold a reference. Otherwise it can happen that you suddenly are holding a wild pointer, that's a pointer to memory that has been freed and potentially allocated again for other stuff.

Using `std::string` is not wrong, but wasteful if you are just handing over string literals.Each and every construction of a string is a memory allocation. In your code snippet, for each call you effectively allocate memory for the file name, copy the file name into it, retrieve the pointer as if it were a string literal and later free the memory again (when the temporary string object goes out of scope). Potentially worse for your path, that is not even a reference, but a string copy. So that would allocate even if `eIconBtz` was a `std::string` already.

Streaming into an uninitialized pointer to a buffer (in your loading function) certainly looks strange. Might be technically OK, an outright access violation or a memory leak depending on the implementation of your ziputils library.

Short version: reading the documentation of used libraries and language features should help a lot with avoiding common mistakes. Even the more boring parts are important.
